File: frontend/src/services/authService.ts
import api from './api';
import { User } from '@/types/user';
import { TokenResponse } from '@/types/api';

export interface LoginPayload {
  email: string;
  password: string;
}

export interface RegisterPayload {
  name: string;
  email: string;
  password: string;
  role?: string;
}

export const authService = {
  async login(payload: LoginPayload): Promise<{ token: string; user: User }> {
    const { data } = await api.post<TokenResponse>('/auth/login', payload);
    localStorage.setItem('access_token', data.access_token);
    const user = await authService.getMe();
    localStorage.setItem('user', JSON.stringify(user));
    return { token: data.access_token, user };
  },

  async register(payload: RegisterPayload): Promise<User> {
    const { data } = await api.post<User>('/auth/register', payload);
    return data;
  },

  async getMe(): Promise<User> {
    const { data } = await api.get<User>('/auth/me');
    return data;
  },

  logout(): void {
    localStorage.removeItem('access_token');
    localStorage.removeItem('user');
  },

  getStoredUser(): User | null {
    const stored = localStorage.getItem('user');
    if (!stored) return null;
    try {
      return JSON.parse(stored) as User;
    } catch {
      return null;
    }
  },

  isAuthenticated(): boolean {
    return !!localStorage.getItem('access_token');
  },
};

File: backend/app/services/auth_service.py
from datetime import timedelta
from sqlalchemy.orm import Session
from app.core.security import verify_password, create_access_token, hash_password
from app.core.config import settings
from app.repositories.user_repository import UserRepository
from app.schemas.auth import LoginRequest, RegisterRequest, TokenResponse
from app.schemas.user import UserResponse
from app.utils.exceptions import UnauthorizedError, ConflictError, BadRequestError
from app.models.user import User


class AuthService:
    def __init__(self, db: Session):
        self.user_repo = UserRepository(db)

    def register(self, data: RegisterRequest) -> UserResponse:
        existing = self.user_repo.get_by_email(data.email)
        if existing:
            raise ConflictError("A user with this email already exists")

        valid_roles = {"admin", "manager", "member"}
        if data.role not in valid_roles:
            raise BadRequestError(f"Role must be one of: {', '.join(valid_roles)}")

        password_hash = hash_password(data.password)
        user = self.user_repo.create(
            name=data.name,
            email=data.email,
            password_hash=password_hash,
            role=data.role,
        )
        return UserResponse.model_validate(user)

    def login(self, data: LoginRequest) -> TokenResponse:
        user = self.user_repo.get_by_email(data.email)
        if not user or not verify_password(data.password, user.password_hash):
            raise UnauthorizedError("Invalid email or password")

        token = create_access_token(
            data={"sub": str(user.id)},
            expires_delta=timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES),
        )
        return TokenResponse(access_token=token)

    def get_me(self, user: User) -> UserResponse:
        return UserResponse.model_validate(user)
